### ISSUE TITLE
iCheck on asp:checkbox prevents autopostback

### ISSUE THREAD

**Issue by YggY** [drgullin/icheck#158]
Hi there

I'm trying to get a pretty checkbox in asp.net to perform a postback method, but unfortunately iCheck prevents my code to execute. Is it possible that iCheck can be modified to handle (a.k.a. not to break) asp postback functions?

Thanks, much appreciated.

Very simple checkbox:
<asp:CheckBox runat="server" CssClass="custom-checkbox" ClientIDMode="Static" ID="cbFilter" OnCheckedChanged="doFilter" AutoPostBack="true"/>

**Comment by Rodbourn** [comment]
Just confirming that using iCheck prevents asp.net from performing postbacks.

**Comment by felipependragon** [comment]
Hi there, 

I was facing this problem as well. I had a page and the postback wasn't firing. So, I did this code below, it saved me, I hope it can help you all.

The code below looks for all ichecks (radiobuttons), with we have some javascript (onclick) it will try to fire it.

```
   //does the iCheck trigger click and fire postback
    $('.icheck').each(function () {
        $onclick = $(this).attr("onclick");
        $iCheckName = $(this).attr("name");
        $buttonTrigger = 'input[name="' + $iCheckName + '"';
        var $this = $(this);
        if ($onclick != undefined) {
            if ($onclick.length > 0) {
                $($buttonTrigger).on('ifChecked', function (event) {
                    $(this).trigger("click");
                });
            }
        }
    });
```

**Comment by joneastwood** [comment]
I also have the same problem - check boxes and/or radio buttons don't auto post back and the above fix didn't work for me

**Comment by felipependragon** [comment]
@joneastwood are you using the iCheck with the class .icheck on them? Could you put a little bit of your code here for us? So we can try to fix it.

**Comment by joneastwood** [comment]
site was built by a third party, it was using iCheck v0.9.1, http://git.io/uhUPMA

CSS looks like this:

.icheckbox_square,

.iradio_square {
    display: block;
    margin: 0;
    padding: 0;
    width: 28px;
    height: 28px;
    background: url(../images/chkbox.png) no-repeat;
    border: none;
    cursor: pointer;
    float:left;}

 .icheckbox_square.checked { background-position: -28px 0;}

.iradio_square {background: url(../images/red_btn.png) no-repeat;}

.iradio_square.checked {background-position: -25px 0;}

Jquery looks like this:

```
    $('input').iCheck({
        checkboxClass: 'icheckbox_square',
        radioClass: 'iradio_square',
    });
```

and my asp.net code looks like this:

```
                    <div class="frm_row">
                        <label class="lbltxt">I would like to</label>
                        <asp:RadioButton ID="RdButEnq" GroupName="Bookeng" CssClass="iradio_square" OnCheckedChanged="schCheckedChanged" AutoPostBack="true" runat="server" Text="Enquire" Checked="true" TextAlign="Right" />
                        <asp:RadioButton ID="RdButBook" GroupName="Bookeng" CssClass="iradio_square" OnCheckedChanged="schCheckedChanged" AutoPostBack="true" runat="server"  Text="Book" TextAlign="Right" />                            
                    </div>
```

basically the autopostback doesnt fire

**Comment by felipependragon** [comment]
Ok. So, I think after rendering the radio button gets the class "iradio_square", doesn't it? If it really happens, we can try this:

at the first line,  
$('.icheck').each(function () { 

change the 
.icheck 

for 
.iradio_square 

on the code snippet

I am not sure if it will help you but may be a solution.

**Comment by joneastwood** [comment]
just tried another page with the latest version and still the same issue

sorry don't follow your solution

where does $('.icheck').each(function () { go and what exactly do i change?

Thanks

**Comment by joneastwood** [comment]
ok code now looks like this:

<script type="text/javascript">
    $(document).ready(function () {
       
        $('input').iCheck({
            checkboxClass: 'icheckbox_minimal-yellow',
            radioClass: 'iradio_minimal-yellow',
            increaseArea: '20%' // optional
        });
        $('.icheckbox_minimal-yellow').each(function () {
            $onclick = $(this).attr("onclick");
            $iCheckName = $(this).attr("name");
            $buttonTrigger = 'input[name="' + $iCheckName + '"';
            var $this = $(this);
            if ($onclick != undefined) {
                if ($onclick.length > 0) {
                    $($buttonTrigger).on('ifChecked', function (event) {
                        $(this).trigger("click");
                    });
                }
            }
        });
        $('.iradio_minimal-yellow').each(function () {
            $onclick = $(this).attr("onclick");
            $iCheckName = $(this).attr("name");
            $buttonTrigger = 'input[name="' + $iCheckName + '"';
            var $this = $(this);
            if ($onclick != undefined) {
                if ($onclick.length > 0) {
                    $($buttonTrigger).on('ifChecked', function (event) {
                        $(this).trigger("click");
                    });
                }
            }
        });
    });
</script>

But still no change still doesn't work and this is from the latest version of Icheck

**Comment by joneastwood** [comment]
```
$(document).ready(function () {
    $('input').iCheck({
        checkboxClass: 'icheckbox_minimal-yellow',
        radioClass: 'iradio_minimal-yellow',
        increaseArea: '20%' // optional
    });


    $('.icheckbox_minimal-yellow').each(function () {
        $onclick = $(this).attr("onclick");
        $iCheckName = $(this).attr("name");
        $buttonTrigger = 'input[name="' + $iCheckName + '"';
        var $this = $(this);
        if ($onclick != undefined) {
            if ($onclick.length > 0) {
                $($buttonTrigger).on('ifChecked', function (event) {
                    $(this).trigger("click");
                });
            }
        }
    });

    $('.iradio_minimal-yellow').each(function () {
        $onclick = $(this).attr("onclick");
        $iCheckName = $(this).attr("name");
        $buttonTrigger = 'input[name="' + $iCheckName + '"';
        var $this = $(this);
        if ($onclick != undefined) {
            if ($onclick.length > 0) {
                $($buttonTrigger).on('ifChecked', function (event) {
                    $(this).trigger("click");
                });
            }
        }
    });

});
```

**Comment by felipependragon** [comment]
@joneastwood I wrote a code here now, summarizing all stuff I used in just one code ... 

<!DOCTYPE html>

<html lang="en" xmlns="http://www.w3.org/1999/xhtml">
<head>
    <meta charset="utf-8" />
    <title></title>
    <style>
        .radio-inline, .checkbox-inline 
        {
            display: inline-block;
            padding-left: 20px;
            margin-bottom: 0;
            vertical-align: middle;
            font-weight: 400;
            cursor: pointer;
        }

    </style>

    <script>//Main Call to your jquery script</script>

    <script>
        //What I've Done
        //First, inside document.ready()
        $(document).ready(function () {

            //It makes all elements which type is input[type=radio] receives the class icheck, and the label for it receives the class radio-inline
            $('input[type=radio]').each(function () {
                var $element = $(this);
                var $label = $("label[for='" + this.id + "']");
                $label.addClass('radio-inline');
                $element.addClass('icheck');
            });

            //This one you have already done using your own styles
            $('.icheck').iCheck({
                checkboxClass: 'icheckbox_minimal-yellow',
                radioClass: 'iradio_minimal-yellow'
            });

            //Here the code snippet will trigger the click
            $('.icheck').each(function () {
                $onclick = $(this).attr("onclick");
                $iCheckName = $(this).attr("name");
                $buttonTrigger = 'input[name="' + $iCheckName + '"';
                var $this = $(this);
                if ($onclick != undefined) {
                    if ($onclick.length > 0) {
                        $($buttonTrigger).on('ifChecked', function (event) {
                            $(this).trigger("click");
                        });
                    }
                }
            });
        });

    </script>
</head>
<body>
    <form id="form1" runat="server">
        <asp:RadioButtonList RepeatDirection="Horizontal" RepeatLayout="Table" ID="rblActive"
            AutoPostBack="True" runat="server" OnSelectedIndexChanged="rblActive_SelectedIndexChanged" runat="server">
            <asp:ListItem Value="1" Text="Active"></asp:ListItem>
            <asp:ListItem Value="0" Text="Not Active"></asp:ListItem>
            <asp:ListItem Value="2" Text="All Of Them" Selected="True"></asp:ListItem>
        </asp:RadioButtonList>
    </form>
</body>
</html>

**Comment by felipependragon** [comment]
```
<style>
    .radio-inline, .checkbox-inline 
    {
        display: inline-block;
        padding-left: 20px;
        margin-bottom: 0;
        vertical-align: middle;
        font-weight: 400;
        cursor: pointer;
    }
</style>

    //What I've Done
    //First, inside document.ready()
    $(document).ready(function () {

        //It makes all elements which type is input[type=radio] receives the class icheck, and the label for it receive the class radio-inline
        $('input[type=radio]').each(function () {
            var $element = $(this);
            var $label = $("label[for='" + this.id + "']");
            $label.addClass('radio-inline');
            $element.addClass('icheck');
        });

        //This one you have already done using your own styles
        $('.icheck').iCheck({
            checkboxClass: 'icheckbox_minimal-yellow',
            radioClass: 'iradio_minimal-yellow'
        });

        //Here the code snippet will trigger the click
        $('.icheck').each(function () {
            $onclick = $(this).attr("onclick");
            $iCheckName = $(this).attr("name");
            $buttonTrigger = 'input[name="' + $iCheckName + '"';
            var $this = $(this);
            if ($onclick != undefined) {
                if ($onclick.length > 0) {
                    $($buttonTrigger).on('ifChecked', function (event) {
                        $(this).trigger("click");
                    });
                }
            }
        });
    });



<form id="form1" runat="server">
    <asp:RadioButtonList RepeatDirection="Horizontal" RepeatLayout="Table" ID="rblActive"
        AutoPostBack="True" runat="server" OnSelectedIndexChanged="rblActive_SelectedIndexChanged" runat="server">
        <asp:ListItem Value="1" Text="Active"></asp:ListItem>
        <asp:ListItem Value="0" Text="Not Active"></asp:ListItem>
        <asp:ListItem Value="2" Text="All Of Them" Selected="True"></asp:ListItem>
    </asp:RadioButtonList>
</form>
```

</body>
</html>

**Comment by felipependragon** [comment]
@joneastwood in this code above I tried to summarize all thing I have done in my project. It works fine for me. 

The entire javascript used above is in a .js file that I've included in masterpage header, so I made it only once and it works for every page created using that masterpage.

**Comment by joneastwood** [comment]
@felipependragon Thanks for all your help, we are getting somewhere

Now my code looks like this:

```
                    <div class="frm_row">
                        <label class="lbltxt">I would like to</label>
                        <asp:RadioButtonList RepeatDirection="Horizontal" RepeatLayout="Table" ID="rblActive"
                            AutoPostBack="True" runat="server" OnSelectedIndexChanged="rblActive_SelectedIndexChanged">
                            <asp:ListItem Value="Enquire" Text="Enquire"></asp:ListItem>
                            <asp:ListItem Value="Book" Text="Book"></asp:ListItem>
                        </asp:RadioButtonList>
                    </div>
```

and

```
                            <asp:CheckBoxList ID="CheckBoxList1"  OnSelectedIndexChanged="CheckBoxList1_SelectedIndexChanged" AutoPostBack="True" runat="server">
                                <asp:ListItem></asp:ListItem>
                            </asp:CheckBoxList>
```

and i have added this to just test things

```
Protected Sub rblActive_SelectedIndexChanged(sender As Object, e As EventArgs) Handles rblActive.SelectedIndexChanged
    MsgBox("radio")
End Sub

Protected Sub CheckBoxList1_SelectedIndexChanged(sender As Object, e As EventArgs) Handles CheckBoxList1.SelectedIndexChanged
    MsgBox("checkbox")
End Sub
```

now all fires but 

I get the msg boxes come up twice no matter if i click a radio button or a checkbox they both pop up twice

**Comment by felipependragon** [comment]
When you click, it fires twice passing through the 2 methods (popping up "radio" and "checkbox") ? I think it could be related to this 

```
    $('.icheck').each(function () {
        $onclick = $(this).attr("onclick");
        $iCheckName = $(this).attr("name");
        $buttonTrigger = 'input[name="' + $iCheckName + '"';
        var $this = $(this);
        if ($onclick != undefined) {
            if ($onclick.length > 0) {
                $($buttonTrigger).on('ifChecked', function (event) {
                    $(this).trigger("click");
                });
            }
        }
    });
```

Because, at that point we are passing through all icheck (it does not matter if it is check or radio) I guess we need to improve it, by now I did not have to put them together, maybe this is the reason why is working fine for me. I will try to simulate something and them if I get some result I'll share with you. ok?

**Comment by joneastwood** [comment]
ok cheers

for now luckily I only need one of the other on the same page to post back, so I have got round it for now, once its all live I will show you the end result :-) - thanks again for your PROMPT help

**Comment by felipependragon** [comment]
@joneastwood No problem :-)  You're Welcome

**Comment by CharlesOkwuagwu** [comment]
a very simple solution:

  $('input').on('ifChecked', function (e) {
                __doPostBack(e.currentTarget, '');
            });

**Comment by amarsalunke** [comment]
icheck plugin with updatepanel in asp.net

**Comment by amarsalunke** [comment]
icheck plugin with update-panel in asp.net not working proper?

**Comment by YggY** [comment]
It does work. Are you using a regular <input> tag or something else?

**Comment by amarsalunke** [comment]
yes i am using a regular tag..can you send sample code Please

**Comment by amarsalunke** [comment]
SelectedIndexChanged and CheckedChanged wokring only first click

**Comment by paoyi** [comment]
Esto me funciono: 

            $('input').iCheck({
                checkboxClass: 'icheckbox_square-green',
                radioClass: 'iradio_square-green',
                increaseArea: '20%' // optional
            });
            $('input').on('ifChecked', function (event) {
                $(this).trigger("click");
            });

**Comment by fevzimormedya** [comment]
I used this for second click:
`$('input').on('ifUnchecked', function (event) {
        $(this).trigger("click");
    });`

**Comment by ghost** [comment]
> a very simple solution:
> 
> $('input').on('ifChecked', function (e) {
> __doPostBack(e.currentTarget, '');
> });

Almost but I gurantee the works

                $('input').on('ifChanged', function (e) {
                    __doPostBack(this.id, '');
                });
